OSSM-8952: Cluster-wide migration using the istio revision label (#89164)
openshift/openshift-docs: modules/ossm-migrating-a-cluster-wide-deployment-using-the-istio-revision-label.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-a-cluster-wide-deployment-using-the-istio-revision-label_{context}"]
= Migrating a cluster-wide deployment using the Istio revision label

This procedure performs a canary upgrade with the gradual migration of data plane namespaces for a cluster-wide deployment using the `istio.io/rev` label.

The `bookinfo` example application is being used for demonstration purposes with a minimal example for the `Istio` resource. For more information on configuration differences between the {SMProduct} 2 `ServiceMeshControlPlane` resource and the {SMProduct} 3 `Istio` resource, see "Configuration fields mapping between Service Mesh 2 and Service Mesh 3".

You can follow these same steps with your own workloads.

.Prerequisites

* You have deployed {ocp-product-title} 4.14 or later.
* You are logged in to the {ocp-product-title} web console as a user with the `cluster-admin` role.
* You have completed the premigration checklists.
* You have the {SMProduct} {SMv2Version} Operator installed.
* You have the {SMProduct} 3 Operator installed.
* You created an `IstioCNI` resource.
* You have the `istioctl` tool installed.

> * You are running a `ClusterWide` `ServiceMeshControlPlane`.

Author: done

* You have installed the `bookinfo` application.

.Procedure

> . Identify the namespaces that contain a 2.6 control plane by running the following command:

Author: @michaelryanpeter I don't disagree; however, we already have published content for related tasks that don't use the term version. In this case, I think it's best for me to discuss this with the Service Mesh doc team and we can consider going back and updated across the doc set. For now, I'd prefer to leave it this way to be consistent  because we have a content deadline looming.

> . Identify the namespaces that contain a 2.6 control plane by running the following command:

Reviewer: SGTM. It isn't something that blocks merging.

+
[source,terminal]
----
$ oc get smcp -A
----
+

> .Sample output:

Author: done

[source,terminal]
----
NAMESPACE      NAME                   READY   STATUS            PROFILES      VERSION   AGE

> istio-system   install-istio-system   6/6     ComponentsReady   ["default"]   2.6.4     115m

Author: done.

----

. Create a YAML file named `ossm-3.yaml` that creates the {istio} resource for the 3.0 installation in the same namespace as the `ServiceMeshControlPlane` resource for the 2.6 installation.
+
.Example configuration
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: ossm-3 <1>
spec:
  updateStrategy:
    type: RevisionBased
  namespace: istio-system <2>

>   version: v1.24.1

Author: done

  values:  <3>
    meshConfig:
      extensionProviders:
        - name: prometheus
          prometheus: {}
        - name: otel
          opentelemetry:
            port: 4317
            service: otel-collector.opentelemetrycollector-3.svc.cluster.local
----

> <1> The `name`, `updateStrategy` and `version` fields are significant for injection labels.

Author: I've tweaked the language and added  a "Fore more information see _____" and included an additional resources section with the link.

> <1> The `name`, `updateStrategy` and `version` fields are significant for injection labels.

Reviewer: Update LGTM! Thanks for clarifying. Note, there is a small typo in the new content. (I left another comment just for belts and suspenders)

<2> The 3.0 and 2.6 control planes must run in the same namespace.

> <3> Optional: If you are migrating metrics and tracing, update the `extensionProviders` fields according to your tracing and metrics configurations.

Author: done

+

> This example configuration does not use any `discoverySelectors`, which means the {istio} control plane has access to all namespaces. If you want to define `discoverySelectors` all data plane namespaces that you plan to migrate from 2.6 must be matched.

Reviewer: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#object-references

Specific API object references must be followed by a noun. Alternatively, you can use a generic reference here and get the added bonus of it reading better:

I also think you are burying the lede here. I don't think the customer wants to define a selector, they either want the control plane to have access to all namespaces or they want to restrict control plane access to specific namespaces. You also jump between talking about the control plane and then data planes. Are these the same? If not, will the customer be able to make that leap with you?

```suggestion
In this example configuration, the {istio} control plane has access to all namespaces on the cluster. If you want limit the namespaces the control plan has access to, you must define discovery selectors. All data plane namespaces that you plan to migrate from version 2.6 must be matched.
```

Also, this might be worth mentioning before the YAML example. If I were a customer, I would want to know if I am granting access to all namespaces before I create the file. WDYT of pulling this up to line 44 and turning it into a NOTE admonition? If you decide to do that, remember to change "this example config" to "the following example config".

Author: Thanks for the suggestion, I moved it up as a note.

. Apply the YAML file by running the following command:
+
[source,terminal]
----
$ oc apply -f ossm-3.yaml
----

. Verify that the new `istiod` resource uses the existing root certificate by running the following command:
+
[source,terminal]
----

> $ oc logs deployments/istiod-ossm-3-v1-24-1 -n istio-system | grep 'Load signing key and cert from existing secret'

Author: done

----
+

> .Sample output:

Reviewer: ```suggestion
.Example output
```

Author: done

istio-system   install-istio-system   6/6     ComponentsReady   ["default"]   2.6.6         115m
----

. Create a YAML file named `ossm-3.yaml` that creates the {istio} resource for the 3.0 installation in the same namespace as the `ServiceMeshControlPlane` resource for the 2.6 installation.
+
[NOTE]
====
In the following example configuration, the {istio} control plane has access to all namespaces on the cluster. If you want limit the namespaces the control plan has access to, you must define discovery selectors. All data plane namespaces that you plan to migrate from version 2.6 must be matched.
====
+
.Example `Istio` resource
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: ossm-3 <1>
spec:
  updateStrategy:
    type: RevisionBased
  namespace: istio-system <2>
  version: v{istio-latest}
  values:  <3>
    meshConfig:
      extensionProviders:
        - name: prometheus
          prometheus: {}
        - name: otel
          opentelemetry:
            port: 4317
            service: otel-collector.opentelemetrycollector-3.svc.cluster.local
----

> <1> The `name`, `updateStrategy` and `version` fields specify how the `IstioRevision` resource name is created. Fore more information, see "Identifying the revision name".

Reviewer: ```suggestion
<1> The `name`, `updateStrategy` and `version` fields specify how the `IstioRevision` resource name is created. For more information, see "Identifying the revision name".
```
Small typo. Change LGTM!

Author: done
